tools/gradle/plugin/src/main/kotlin/com/zeroc/ice/SliceToolsResourceExtractor.kt

```diff
@@ -0,0 +1,118 @@
+                gradleUserDir.mkdirs()
+                Files.copy(resourceStream, slice2JavaFile.toPath(), StandardCopyOption.REPLACE_EXISTING)
+                slice2JavaFile.setExecutable(true)
+            } else {
+                return null
+            }
+        }
+
+        return gradleUserDir.absolutePath
+    }
+
+    /**
+     * Extracts the Slice files from the plugin resources to the Gradle user home directory.
+     *
+     * @param project The Gradle project
+     * @return The path to the directory containing the Slice files, or null if the extraction failed or the
+     * Slice files are not present in the resources.
+     */
+    fun extractSliceFiles(project: Project): String? {
+        val pluginVersion = getPluginVersion(project)
+
+        val sliceSourcePath = "/slice/"
+        val sliceTargetDir = File(project.gradle.gradleUserHomeDir, "native/$pluginName-$pluginVersion/slice")
+
+        if (!sliceTargetDir.exists()) {
+            val resourceURL = this::class.java.getResource(sliceSourcePath)
+            if (resourceURL == null) {
+                return null
+            }
+
+            project.logger.lifecycle("Extracting Slice files to ${sliceTargetDir.absolutePath}")
+            sliceTargetDir.mkdirs()
+
+            val resourceDir = File(resourceURL.toURI())
+            copyDirectory(resourceDir, sliceTargetDir)
+        }
+
+        return sliceTargetDir.absolutePath
+    }
+
+    private fun getPluginVersion(project: Project): String {
+        val pkg = this::class.java.`package`
+        val implementationVersion = pkg?.implementationVersion
+        val projectSliceToolsVersion = project.findProperty("sliceToolsVersion") as? String
+        return implementationVersion
+            ?: projectSliceToolsVersion
+            ?: throw IllegalStateException("Failed to determine plugin version: implementationVersion is missing")
+    }
+
+    private fun copyDirectory(source: File, destination: File) {
+        source.walk().forEach { file ->
+            val targetFile = File(destination, file.relativeTo(source).path)
+            if (file.isDirectory) {
+                targetFile.mkdirs()
+            } else {
+                file.inputStream().use { input ->
+                    targetFile.outputStream().use { output -> input.copyTo(output) }
+                }
+            }
+        }
+    }
+
+    private fun detectOS(): String {
+        val os = System.getProperty("os.name").lowercase()
+        return when {
+            os.contains("win") -> "windows"
+            os.contains("mac") -> "macos"
+            os.contains("nix") || os.contains("nux") -> "linux"
+            else -> os
+        }
+    }
+
+    private fun detectArch(): String {
+        val arch = System.getProperty("os.arch").lowercase()
+        return when {
+            arch.contains("64") || arch.contains("x86_64") -> "x64"
+            arch.contains("arm") || arch.contains("aarch64") -> "arm64"
```

Comment:
Should we just bundle the compilers in directories that match what Kotlin uses for os.arch? Or would that just be too painful.

Reply:
Not sure, here I just stick which what we do for other builds, this mapping of names is simple enough and keep same layout for all tools.